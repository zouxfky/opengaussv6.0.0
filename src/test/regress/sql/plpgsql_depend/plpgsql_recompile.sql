drop schema if exists plpgsql_recompile cascade;
create schema plpgsql_recompile;
set current_schema = plpgsql_recompile;

create type s_type as (
    id integer,
    name varchar,
    addr text
);

--- 不设置 plpgsql_dependency
set behavior_compat_options = '';
--- test 0
create or replace procedure proc_no_guc(a s_type)
is
begin
RAISE INFO 'call a: %', a;
end;
/
call proc_no_guc(((1,'zhang','shanghai')));

alter type s_type ADD attribute a int;
--- 修改后，预期新增的参数值不打印
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
call proc_no_guc((1,'zhang','shanghai',100));
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));

alter type s_type DROP attribute a;
--- 修改后，预期valid字段无变化
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
call proc_no_guc((1,'zhang','shanghai'));
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
--- drop procedure
drop procedure proc_no_guc;

--- 设置 plpgsql_dependency
set behavior_compat_options = 'plpgsql_dependency';
--- test 1
create or replace procedure type_alter(a s_type)
is
begin    
    RAISE INFO 'call a: %', a;
end;
/
select valid from pg_object where object_type='P' and object_oid 
in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = 
(select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
call type_alter((1,'zhang','shanghai'));

alter function type_alter compile;
alter procedure type_alter(s_type) compile;

alter type s_type ADD attribute a int;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
call type_alter((1,'zhang','shanghai',100));
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));

alter type s_type DROP attribute a;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
call type_alter((1,'zhang','shanghai'));
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));

alter type s_type ALTER attribute id type float;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
alter procedure type_alter compile;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
call type_alter((1.0,'zhang','shanghai'));
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));

-- report error
alter type s_type RENAME to s_type_rename;
ALTER type s_type RENAME ATTRIBUTE name TO new_name;
call type_alter((1,'zhang','shanghai'));

-- test 2
drop table if exists stu;
create table stu(sno int, name varchar, sex varchar, cno int);
create type r1 as (a int, c stu%RowType);

drop package if exists pkg;
create or replace package pkg
is    
procedure proc1(p_in r1);
end pkg;
/
create or replace package body pkg
is
declare
v1 r1;
v2 stu%RowType;
procedure proc1(p_in r1) as
begin        
RAISE INFO 'call p_in: %', p_in;
end;
end pkg;
/

call pkg.proc1((1,(1,'zhang','M',1)));
alter package pkg compile;
alter package pkg compile specification;
alter package pkg compile body;
alter package pkg compile package;

alter table stu ADD column b int;
alter package pkg compile;
call pkg.proc1((1,(1,'zhang','M',1,2)));


alter table stu DROP column b;
alter package pkg compile;
call pkg.proc1((1,(1,'zhang','M',1)));

alter table stu RENAME to stu_re;
alter package pkg compile;
call pkg.proc1((1,(1,'zhang','M',1)));

alter type r1 ALTER attribute a type float;
alter package pkg compile;
call pkg.proc1((1.0,(1,'zhang','M',1)));

--compile error
create or replace package body pkg
is
declare
v1 r1%RowType;
procedure proc1(p_in r1) as
begin
select into v1 values (1,ROW(1,'zhang','M',1));     
RAISE INFO 'call p_in: %', p_in;
end;
end pkg;
/
alter package pkg compile;

-- test 3
create procedure test_subtype_proc is
xx s_type;
begin
xx := (1,'aaa','bbb');
raise notice '%',xx;
end;
/

call test_subtype_proc();
alter type s_type ADD attribute a int;
--修改后，预期调用成功
call test_subtype_proc();
--手动触发编译， 调用成功
alter procedure test_subtype_proc compile;
call test_subtype_proc();

-- test 4
create or replace package pkg
is    
procedure proc1();
end pkg;
/
create or replace package body pkg
is
xxx s_type;
procedure proc1() as
yyy s_type;
begin        
xxx:= (1,'aaa','aaaaa');
yyy:= (2,'bbb','bbbbb');
RAISE INFO 'call xxx: %, yyy: %', xxx, yyy;
end;
end pkg;
/

call pkg.proc1();
alter type s_type ADD attribute b int;
--修改后，预期调用成功
call pkg.proc1();
--手动触发编译， 调用成功
alter package pkg compile;
call pkg.proc1();

-- test 5
create or replace package pkg
is    
procedure proc1(p_in s_type);
end pkg;
/
create or replace package body pkg
is
xxx s_type;
procedure proc1(p_in s_type) as
yyy s_type;
begin        
xxx:= (1,'aaa','aaaaa');
yyy:= (2,'bbb','bbbbb');
RAISE INFO 'call xxx: %, yyy: %, p_in: %', xxx, yyy, p_in;
end;
end pkg;
/

call pkg.proc1((1,'zhang','M',1,2));
alter type s_type ADD attribute c int;
--修改后，预期调用成功
call pkg.proc1((1,'zhang','M',1,2,3));
--手动触发编译， 调用成功
alter package pkg compile;
call pkg.proc1((1,'zhang','M',1,2,3));

-- test 6
drop type if exists r1;
drop table if exists stu;
create table stu(sno int, name varchar, sex varchar, cno int);
create type r1 as (a int, c stu%RowType);
create schema plpgsql_recompile_new;

create or replace procedure test_proc(p_in r1)
is
declare
v1 stu%RowType;
begin        
RAISE INFO 'call p_in: %', p_in;
end;
/
call test_proc((1,(1,'zhang','M',1)));

-- rename report error
ALTER TYPE r1 RENAME to r1_rename;
ALTER TYPE r1 RENAME ATTRIBUTE a TO new_a;
ALTER TABLE stu RENAME  TO stu_rename;
ALTER TABLE stu RENAME COLUMN cno TO new_cno;
ALTER PROCEDURE test_proc(r1) RENAME TO test_proc_rename;
ALTER SCHEMA plpgsql_recompile RENAME TO plpgsql_recompile_rename;
call test_proc((1.0,(1,'zhang','M',1)));

-- set schema report error
ALTER TYPE r1 SET SCHEMA plpgsql_recompile_new;
ALTER TABLE stu SET SCHEMA plpgsql_recompile_new;
ALTER PROCEDURE test_proc(r1) SET SCHEMA plpgsql_recompile_new;
call test_proc((1.0,(1,'zhang','M',1)));

-- drop and recreate
drop type r1;
drop table stu;
drop procedure test_proc;

create table stu(sno int, name varchar, sex varchar, cno int);
create type r1 as (a int, c stu%RowType);
create or replace procedure test_proc(p_in r1)
is
declare
v1 stu%RowType;
begin        
RAISE INFO 'call p_in: %', p_in;
end;
/
call test_proc((1,(1,'zhang','M',1)));

drop type r1;
drop table stu;
drop procedure test_proc;

-- test 7
set behavior_compat_options = 'plpgsql_dependency';
create or replace package pkg1 is
type tttt is record (col1 int, col2 text);
procedure p1(param pkg2.tqqq);
end pkg1;
/
create or replace package pkg2 is
type tqqq is record (col1 int, col2 text, col3 varchar);
procedure p1(param pkg1.tttt);
end pkg2;
/
create or replace package body pkg2 is
procedure p1(param pkg1.tttt) is
begin
RAISE INFO 'call param: %', param;
end;
end pkg2;
/

call pkg2.p1((1,'a'));
set behavior_compat_options ='';
drop package pkg1;
set behavior_compat_options = 'plpgsql_dependency';
drop package pkg2;

-- clean
drop schema plpgsql_recompile_new cascade;
drop schema plpgsql_recompile cascade;
reset behavior_compat_options;
