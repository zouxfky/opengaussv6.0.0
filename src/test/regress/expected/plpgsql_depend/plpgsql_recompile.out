drop schema if exists plpgsql_recompile cascade;
NOTICE:  schema "plpgsql_recompile" does not exist, skipping
create schema plpgsql_recompile;
set current_schema = plpgsql_recompile;
create type s_type as (
    id integer,
    name varchar,
    addr text
);
--- 不设置 plpgsql_dependency
set behavior_compat_options = '';
--- test 0
create or replace procedure proc_no_guc(a s_type)
is
begin
RAISE INFO 'call a: %', a;
end;
/
call proc_no_guc(((1,'zhang','shanghai')));
INFO:  call a: (1,zhang,shanghai)
 proc_no_guc 
-------------
 
(1 row)

alter type s_type ADD attribute a int;
--- 修改后，预期新增的参数值不打印
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

call proc_no_guc((1,'zhang','shanghai',100));
INFO:  call a: (1,zhang,shanghai,)
 proc_no_guc 
-------------
 
(1 row)

select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

alter type s_type DROP attribute a;
--- 修改后，预期valid字段无变化
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

call proc_no_guc((1,'zhang','shanghai'));
INFO:  call a: (1,zhang,shanghai)
 proc_no_guc 
-------------
 
(1 row)

select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='proc_no_guc' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

--- drop procedure
drop procedure proc_no_guc;
--- 设置 plpgsql_dependency
set behavior_compat_options = 'plpgsql_dependency';
--- test 1
create or replace procedure type_alter(a s_type)
is
begin    
    RAISE INFO 'call a: %', a;
end;
/
select valid from pg_object where object_type='P' and object_oid 
in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = 
(select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

call type_alter((1,'zhang','shanghai'));
INFO:  call a: (1,zhang,shanghai)
 type_alter 
------------
 
(1 row)

alter function type_alter compile;
WARNING:  Functions type_alter does not exist, if it is a stored procedure, use ALTER PROCEDURE.
alter procedure type_alter(s_type) compile;
alter type s_type ADD attribute a int;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 f
(1 row)

call type_alter((1,'zhang','shanghai',100));
INFO:  call a: (1,zhang,shanghai,100)
 type_alter 
------------
 
(1 row)

select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

alter type s_type DROP attribute a;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 f
(1 row)

call type_alter((1,'zhang','shanghai'));
INFO:  call a: (1,zhang,shanghai)
 type_alter 
------------
 
(1 row)

select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

alter type s_type ALTER attribute id type float;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 f
(1 row)

alter procedure type_alter compile;
select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

call type_alter((1.0,'zhang','shanghai'));
INFO:  call a: (1,zhang,shanghai)
 type_alter 
------------
 
(1 row)

select valid from pg_object where object_type='P' and object_oid in (select Oid from pg_proc where propackageid = 0 and proname='type_alter' and pronamespace = (select Oid from pg_namespace where nspname = 'plpgsql_recompile'));
 valid 
-------
 t
(1 row)

-- report error
alter type s_type RENAME to s_type_rename;
ERROR:  The rename operator of s_type is not allowed, because it is referenced by the other object.
ALTER type s_type RENAME ATTRIBUTE name TO new_name;
ERROR:  cannot rename attribute of the type because it is dependent on another object.
call type_alter((1,'zhang','shanghai'));
INFO:  call a: (1,zhang,shanghai)
 type_alter 
------------
 
(1 row)

-- test 2
drop table if exists stu;
NOTICE:  table "stu" does not exist, skipping
create table stu(sno int, name varchar, sex varchar, cno int);
create type r1 as (a int, c stu%RowType);
drop package if exists pkg;
NOTICE:  package pkg() does not exist, skipping
create or replace package pkg
is    
procedure proc1(p_in r1);
end pkg;
/
create or replace package body pkg
is
declare
v1 r1;
v2 stu%RowType;
procedure proc1(p_in r1) as
begin        
RAISE INFO 'call p_in: %', p_in;
end;
end pkg;
/
call pkg.proc1((1,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 proc1 
-------
 
(1 row)

alter package pkg compile;
alter package pkg compile specification;
alter package pkg compile body;
alter package pkg compile package;
alter table stu ADD column b int;
alter package pkg compile;
call pkg.proc1((1,(1,'zhang','M',1,2)));
INFO:  call p_in: (1,"(1,zhang,M,1,2)")
 proc1 
-------
 
(1 row)

alter table stu DROP column b;
alter package pkg compile;
call pkg.proc1((1,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 proc1 
-------
 
(1 row)

alter table stu RENAME to stu_re;
ERROR:  The rename operator on stu is not allowed, because it is dependent on another object.
alter package pkg compile;
call pkg.proc1((1,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 proc1 
-------
 
(1 row)

alter type r1 ALTER attribute a type float;
alter package pkg compile;
call pkg.proc1((1.0,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 proc1 
-------
 
(1 row)

--compile error
create or replace package body pkg
is
declare
v1 r1%RowType;
procedure proc1(p_in r1) as
begin
select into v1 values (1,ROW(1,'zhang','M',1));     
RAISE INFO 'call p_in: %', p_in;
end;
end pkg;
/
WARNING:  syntax error at or near "(" when compile function proc1(r1)
DETAIL:  N/A
CONTEXT:  compilation of PL/pgSQL package near line 1
alter package pkg compile;
ERROR:  syntax error at or near "(" when compile function proc1(r1)
LINE 2: select into v1 values (1,ROW(1,'zhang','M',1));     
        ^
DETAIL:  syntax error
QUERY:   DECLARE begin
select into v1 values (1,ROW(1,'zhang','M',1));     
RAISE INFO 'call p_in: %', p_in;
end
-- test 3
create procedure test_subtype_proc is
xx s_type;
begin
xx := (1,'aaa','bbb');
raise notice '%',xx;
end;
/
call test_subtype_proc();
NOTICE:  (1,aaa,bbb)
 test_subtype_proc 
-------------------
 
(1 row)

alter type s_type ADD attribute a int;
--修改后，预期调用成功
call test_subtype_proc();
NOTICE:  (1,aaa,bbb,)
 test_subtype_proc 
-------------------
 
(1 row)

--手动触发编译， 调用成功
alter procedure test_subtype_proc compile;
call test_subtype_proc();
NOTICE:  (1,aaa,bbb,)
 test_subtype_proc 
-------------------
 
(1 row)

-- test 4
create or replace package pkg
is    
procedure proc1();
end pkg;
/
create or replace package body pkg
is
xxx s_type;
procedure proc1() as
yyy s_type;
begin        
xxx:= (1,'aaa','aaaaa');
yyy:= (2,'bbb','bbbbb');
RAISE INFO 'call xxx: %, yyy: %', xxx, yyy;
end;
end pkg;
/
call pkg.proc1();
INFO:  call xxx: (1,aaa,aaaaa,), yyy: (2,bbb,bbbbb,)
 proc1 
-------
 
(1 row)

alter type s_type ADD attribute b int;
--修改后，预期调用成功
call pkg.proc1();
INFO:  call xxx: (1,aaa,aaaaa,,), yyy: (2,bbb,bbbbb,,)
 proc1 
-------
 
(1 row)

--手动触发编译， 调用成功
alter package pkg compile;
call pkg.proc1();
INFO:  call xxx: (1,aaa,aaaaa,,), yyy: (2,bbb,bbbbb,,)
 proc1 
-------
 
(1 row)

-- test 5
create or replace package pkg
is    
procedure proc1(p_in s_type);
end pkg;
/
create or replace package body pkg
is
xxx s_type;
procedure proc1(p_in s_type) as
yyy s_type;
begin        
xxx:= (1,'aaa','aaaaa');
yyy:= (2,'bbb','bbbbb');
RAISE INFO 'call xxx: %, yyy: %, p_in: %', xxx, yyy, p_in;
end;
end pkg;
/
call pkg.proc1((1,'zhang','M',1,2));
INFO:  call xxx: (1,aaa,aaaaa,,), yyy: (2,bbb,bbbbb,,), p_in: (1,zhang,M,1,2)
 proc1 
-------
 
(1 row)

alter type s_type ADD attribute c int;
--修改后，预期调用成功
call pkg.proc1((1,'zhang','M',1,2,3));
INFO:  call xxx: (1,aaa,aaaaa,,,), yyy: (2,bbb,bbbbb,,,), p_in: (1,zhang,M,1,2,3)
 proc1 
-------
 
(1 row)

--手动触发编译， 调用成功
alter package pkg compile;
call pkg.proc1((1,'zhang','M',1,2,3));
INFO:  call xxx: (1,aaa,aaaaa,,,), yyy: (2,bbb,bbbbb,,,), p_in: (1,zhang,M,1,2,3)
 proc1 
-------
 
(1 row)

-- test 6
drop type if exists r1;
drop table if exists stu;
create table stu(sno int, name varchar, sex varchar, cno int);
create type r1 as (a int, c stu%RowType);
create schema plpgsql_recompile_new;
create or replace procedure test_proc(p_in r1)
is
declare
v1 stu%RowType;
begin        
RAISE INFO 'call p_in: %', p_in;
end;
/
call test_proc((1,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 test_proc 
-----------
 
(1 row)

-- rename report error
ALTER TYPE r1 RENAME to r1_rename;
ERROR:  The rename operator of r1 is not allowed, because it is referenced by the other object.
ALTER TYPE r1 RENAME ATTRIBUTE a TO new_a;
ERROR:  cannot rename attribute of the type because it is dependent on another object.
ALTER TABLE stu RENAME  TO stu_rename;
ERROR:  The rename operator on stu is not allowed, because it is dependent on another object.
ALTER TABLE stu RENAME COLUMN cno TO new_cno;
ERROR:  cannot rename attribute of the type because it is dependent on another object.
ALTER PROCEDURE test_proc(r1) RENAME TO test_proc_rename;
ERROR:  The rename operator of test_proc is not allowed, because it is referenced by the other object.
ALTER SCHEMA plpgsql_recompile RENAME TO plpgsql_recompile_rename;
ERROR:  The rename operator of plpgsql_recompile is not allowed, because it is referenced by the other object.
call test_proc((1.0,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 test_proc 
-----------
 
(1 row)

-- set schema report error
ALTER TYPE r1 SET SCHEMA plpgsql_recompile_new;
ERROR:  The set schema operator of r1 is not allowed, because it is referenced by the other object.
ALTER TABLE stu SET SCHEMA plpgsql_recompile_new;
ERROR:  The set schema operator of stu is not allowed, because it is referenced by another object.
ALTER PROCEDURE test_proc(r1) SET SCHEMA plpgsql_recompile_new;
ERROR:  The set schema operator of test_proc is not allowed, because it is referenced by the other object.
call test_proc((1.0,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 test_proc 
-----------
 
(1 row)

-- drop and recreate
drop type r1;
drop table stu;
drop procedure test_proc;
create table stu(sno int, name varchar, sex varchar, cno int);
create type r1 as (a int, c stu%RowType);
create or replace procedure test_proc(p_in r1)
is
declare
v1 stu%RowType;
begin        
RAISE INFO 'call p_in: %', p_in;
end;
/
call test_proc((1,(1,'zhang','M',1)));
INFO:  call p_in: (1,"(1,zhang,M,1)")
 test_proc 
-----------
 
(1 row)

drop type r1;
drop table stu;
drop procedure test_proc;
-- test 7
set behavior_compat_options = 'plpgsql_dependency';
create or replace package pkg1 is
type tttt is record (col1 int, col2 text);
procedure p1(param pkg2.tqqq);
end pkg1;
/
WARNING:  Type tqqq does not exist.
WARNING:  The header information of function p1 is not defined.
CONTEXT:  compilation of PL/pgSQL package near line 1
WARNING:  Package created with compilation erors.
create or replace package pkg2 is
type tqqq is record (col1 int, col2 text, col3 varchar);
procedure p1(param pkg1.tttt);
end pkg2;
/
--?.*
WARNING:  Package created with compilation erors.
create or replace package body pkg2 is
procedure p1(param pkg1.tttt) is
begin
RAISE INFO 'call param: %', param;
end;
end pkg2;
/
--?.*
WARNING:  Package Body created with compilation erors.
call pkg2.p1((1,'a'));
INFO:  call param: (1,a)
 p1 
----
 
(1 row)

set behavior_compat_options ='';
drop package pkg1;
--?.*
set behavior_compat_options = 'plpgsql_dependency';
drop package pkg2;
NOTICE:  drop cascades to function p1
WARNING:  The oid of the input parameter type of function p1 does not exist.
WARNING:  function p1 does not exist
WARNING:  function p1(plpgsql_recompile.pkg2.tqqq) does not exist
WARNING:  function p1 does not exist
WARNING:  function p1(plpgsql_recompile.pkg2.tqqq) does not exist
-- clean
drop schema plpgsql_recompile_new cascade;
drop schema plpgsql_recompile cascade;
--?.*
DETAIL:  drop cascades to type s_type
drop cascades to function type_alter(s_type)
drop cascades to function test_subtype_proc()
--?.*
drop cascades to function plpgsql_recompile.proc1(s_type)
reset behavior_compat_options;
